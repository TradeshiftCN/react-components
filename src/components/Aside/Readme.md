### Basic Usage

```jsx
import { Button } from '@tradeshift/react-components';

initialState = {
	visible: false
};

<div>
	<Button type="primary" onClick={() => setState({ visible: true })}>
		Open
	</Button>
	<Aside
		visible={state.visible}
		onClose={() => setState({ visible: false })}
		title="Aside"
		style={{ whiteSpace: 'pre' }}
	>
		{new Array(100).fill('row\n\r')}
	</Aside>
</div>;
```
