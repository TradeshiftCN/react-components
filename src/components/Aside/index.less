@import '../../style/variables/index';

@keyframes tsrcDrawerFadeIn {
	0% {
		opacity: 0;
	}
	100% {
		opacity: 1;
	}
}

.ts-rc-aside {
	@duration: 0.3s;
	@timing-function: cubic-bezier(0.7, 0.2, 0.1, 1);
	position: fixed;
	z-index: 1999;
	width: 0%;
	height: 100%;
	transition: transform @duration @timing-function, height 0s ease @duration,
		width 0s ease @duration;
	& > * {
		transition: transform @duration @timing-function, box-shadow @duration @timing-function;
	}
	&-content-wrapper {
		box-shadow: 0 0 3px rgba(0, 0, 0, 0.5);
		position: absolute;
		height: 100%;
	}
	&-content {
		width: 100%;
		height: 100%;
	}
	&-left,
	&-right {
		top: 0;
	}
	&-right {
		right: 0;
		.ts-rc-aside-content-wrapper {
			right: 0;
		}
	}
	&-left {
		left: 0;
		.ts-rc-aside-content-wrapper {
			left: 0;
		}
	}
	&-open {
		width: 100%;
		transition: transform @duration @timing-function;
		.ts-rc-aside-mask {
			height: 100%;
			opacity: 1;
			transition: none;
			animation: tsrcDrawerFadeIn @duration @timing-function;
		}
	}
	&-mask {
		position: absolute;
		top: 0;
		left: 0;
		width: 100%;
		height: 0;
		background-color: rgba(0, 0, 0, 0.71);
		opacity: 0;
		filter: ~'alpha(opacity=45)';
		transition: opacity @duration linear, height 0s ease @duration;
	}
	&-content {
		width: 100%;
		height: 100%;
		background-color: @ts-color-white;
		display: flex;
		flex-direction: column;
	}
	&-header {
		border-bottom: 1px solid @ts-color-gray-light;
		padding-left: @ts-unit;
		padding-right: @ts-unit;
		height: @ts-unit * 3 + 1px;
		line-height: @ts-unit * 3;
		position: relative;
		font-size: @ts-fontsize-big;
	}
	&-close-btn {
		width: @ts-unit-double;
		height: @ts-unit-double;
		line-height: @ts-unit-double;
		border-radius: 50%;
		position: absolute;
		right: @ts-unit-half;
		top: @ts-unit-half;
		color: inherit;
		font-size: 0px;
		text-decoration: none !important;
		transition: 0.2s;
		&:hover {
			background-color: @ts-color-gray-lighter;
		}
		i {
			vertical-align: middle;
		}
	}
	&-body {
		padding: @ts-unit;
		flex: 1;
		overflow: auto;
	}
	&--closable {
		.ts-rc-aside-header {
			padding-right: 55px;
		}
	}
}
