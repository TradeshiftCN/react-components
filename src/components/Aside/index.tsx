import React from 'react';
import RcDrawer from 'rc-drawer';
import PropTypes from 'prop-types';
import cx from 'classnames';
import _ from 'lodash';

type AsideParent = React.RefObject<Aside> | Array<React.RefObject<Aside>>;

export type AsideProps = {
	className?: string;
	children?: React.ReactNode;
	visible?: boolean;
	width?: number;
	placement?: 'left' | 'right';
	showMask?: boolean;
	maskClosable?: boolean;
	title?: React.ReactNode;
	style?: React.CSSProperties;
	closable?: boolean;
	onClose?(): void;
	parent?: AsideParent;
	onChange?(open?: boolean): void;
};

const prefixCls = 'ts-rc-aside';

class Aside extends React.Component<AsideProps> {
	static propTypes = {
		className: PropTypes.string,
		children: PropTypes.node,
		visible: PropTypes.bool,
		width: PropTypes.number,
		placement: PropTypes.oneOf(['left', 'right']),
		showMask: PropTypes.bool,
		maskClosable: PropTypes.bool,
		title: PropTypes.node,
		style: PropTypes.object,
		closable: PropTypes.bool,
		onClose: PropTypes.func,
		parent: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
		onChange: PropTypes.func
	};
	static defaultProps = {
		width: 320,
		placement: 'right',
		maskClosable: true,
		showMask: true,
		closable: true
	};
	public asideChildren: Aside[] = [];

	render() {
		const { visible, children, title, closable, className, parent, ...drawerProps } = this.props;
		return (
			<RcDrawer
				{...drawerProps}
				className={cx(className, { [`${prefixCls}--closable`]: closable })}
				prefixCls={prefixCls}
				open={this.asideChildren.some(v => v.props.visible) ? false : visible}
				handler={false}
				level={null}
			>
				<header className={`${prefixCls}-header`}>
					{title}
					{closable && (
						<button
							type="button"
							className={`${prefixCls}-close-btn`}
							onClick={() => {
								if (this.props.onClose) {
									this.props.onClose();
								}
							}}
						>
							<i className="ts-icon-close"></i>
						</button>
					)}
				</header>
				<div className={`${prefixCls}-body`}>{children}</div>
			</RcDrawer>
		);
	}

	componentDidMount() {
		this.updateParent(this.props.parent, undefined);
	}

	componentDidUpdate(prevProps: AsideProps) {
		this.updateParent(this.props.parent, prevProps.parent);
		if (this.props.visible !== prevProps.visible) {
			setTimeout(() => {
				this.getParents(this.props.parent).forEach(aside => {
					aside.forceUpdate();
				});
			});
		}
	}

	componentWillUnmount() {
		this.updateParent(undefined, this.props.parent);
	}

	private getParents(parent?: AsideParent) {
		if (!parent) {
			return [];
		}
		const parents = Array.isArray(parent) ? parent.map(v => v.current) : [parent.current];
		return _.compact(parents);
	}

	private updateParent(parent?: AsideParent, prevParent?: AsideParent) {
		if (parent === prevParent) {
			return;
		}
		const parents = this.getParents(parent);
		const prevParents = this.getParents(prevParent);
		const increase = _.difference(parents, prevParents);
		const decrease = _.difference(prevParents, parents);

		increase.forEach(aside => {
			if (!aside.asideChildren.includes(this)) {
				aside.asideChildren.push(this);
			}
		});
		decrease.forEach(aside => {
			const index = aside.asideChildren.indexOf(this);
			if (index >= 0) {
				aside.asideChildren.splice(index, 1);
				aside.forceUpdate();
			}
		});
	}
}

export default Aside;
