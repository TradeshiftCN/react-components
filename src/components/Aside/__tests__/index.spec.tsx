import React from 'react';
import { shallow, mount } from 'enzyme';
import Aside from '..';

describe('Aside', () => {
	it('render', () => {
		const wrapper = shallow(<Aside />);
		expect(wrapper).toMatchSnapshot();
	});

	it('parent', () => {
		const ref = React.createRef<Aside>();
		const parent = mount(<Aside ref={ref} visible={true} />);
		const child = mount(<Aside parent={ref} visible={true} />);
		const parentInstance = parent.instance() as Aside;
		expect(parentInstance.asideChildren.length).toBe(1);
		child.unmount();
		expect(parentInstance.asideChildren.length).toBe(0);
	});

	it('closable', () => {
		const onClose = jest.fn();
		const wrapper = shallow(<Aside visible={true} onClose={onClose} />);
		wrapper.find('button').simulate('click');
		expect(onClose).toHaveBeenCalled();
		wrapper.setProps({ closable: false });
		expect(wrapper.find('button').exists()).toBe(false);
	});
});
