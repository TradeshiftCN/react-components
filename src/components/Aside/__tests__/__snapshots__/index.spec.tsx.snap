// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Aside render 1`] = `
<DrawerWrapper
  afterVisibleChange={[Function]}
  className="ts-rc-aside--closable"
  defaultOpen={false}
  duration=".3s"
  ease="cubic-bezier(0.78, 0.14, 0.15, 0.86)"
  forceRender={false}
  getContainer="body"
  handler={false}
  keyboard={true}
  level={null}
  maskClosable={true}
  maskStyle={Object {}}
  onChange={[Function]}
  placement="right"
  prefixCls="ts-rc-aside"
  showMask={true}
  width={320}
  wrapperClassName=""
>
  <header
    className="ts-rc-aside-header"
  >
    <button
      className="ts-rc-aside-close-btn"
      onClick={[Function]}
      type="button"
    >
      <i
        className="ts-icon-close"
      />
    </button>
  </header>
  <div
    className="ts-rc-aside-body"
  />
</DrawerWrapper>
`;
